File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}

        for char in s:
            if char in mapping.values():  # opening brackets
                stack.append(char)
            elif char in mapping:  # closing brackets
                if not stack or mapping[char] != stack.pop():
                    return False
            else:
                return False  # invalid character
        return not stack  # stack should be empty if valid

File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
    const stack = [];
    const mapping = {
        ')': '(',
        '}': '{',
        ']': '['
    };

    for (let char of s) {
        if (Object.values(mapping).includes(char)) {
            stack.push(char);
        } else if (mapping[char]) {
            if (stack.length === 0 || stack.pop() !== mapping[char]) {
                return false;
            }
        } else {
            return false; // invalid character
        }
    }

    return stack.length === 0;
}